File: src/CSharp/501-600/599.MinimumIndexSumOfTwoLists.cs


using System;
using System.Linq;
using System.Collections.Generic;

namespace PeefyLeetCode.MinimumIndexSumOfTwoLists
{
    public class Solution
    {
        public string[] FindRestaurant(string[] list1, string[] list2)
        {
            int m = list1.Length;
            int n = list2.Length;
            int min = m + n;
            var hash = new Dictionary<string, int>();
            var res = new List<string>();
            for (int i = 0; i < n; i++)
            {
                hash.Add(list2[i], i);
            }
            for (int i = 0; i < m; i++)
            {
                if (hash.ContainsKey(list1[i]))
                {
                    if (min > i + hash[list1[i]])
                    {
                        res.Clear();
                        res.Add(list1[i]);
                        min = i + hash[list1[i]];
                    }
                    else if (min == i + hash[list1[i]])
                    {
                        res.Add(list1[i]);
                    }
                }
            }
            return res.ToArray();
        }
    }
}

File: src/Java/599.MinimumIndexSumOfTwoLists.java

import java.awt.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.Queue;
import java.util.HashMap;
import java.util.HashSet;

class Solution {
    public String[] findRestaurant(String[] list1, String[] list2) {
        int m = list1.length;
        int n = list2.length;
        int min = m + n;
        HashMap<String, Integer> hash = new HashMap<>();
        LinkedList<String> res = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            hash.put(list2[i], i);
        }
        for (int i = 0; i < m; i++) {
            if (hash.containsKey(list1[i])) {
                if (min > i + hash.get(list1[i])) {
                    res.clear();
                    res.add(list1[i]);
                    min = i + hash.get(list1[i]);
                }
                else if(min == i + hash.get(list1[i])){
                    res.add(list1[i]);
                }
            }
        }
        return res.toArray(new String[res.size()]);
    }
}

